8330164

xmas92 opened the pull request:
The verification added in [JDK-8329757](https://bugs.openjdk.org/browse/JDK-8329757) will not work then deoptimization occurs on a monitorenter bytecode. The locking may be in a transitional state. This patch will skip the verification when this occurs.

dean-long commented:
Can you describe the transitional state this fix avoids, and why it only needs to deal with monitorenter from a synchronized method prologue, and not also monitorenter from synchronized blocks.

dean-long commented:
> Can you describe the transitional state this fix avoids, and why it only needs to deal with monitorenter from a synchronized method prologue, and not also monitorenter from synchronized blocks.

Nevermind, I see that it does handle monitorenter for synchronized methods.

dean-long commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -440,8 +440,15 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
+    // If deoptimizing from monitorenter bytecode we maybe in transitional state. Skip verification.
+    if (!is_syncronized_entry && bc != Bytecodes::Code::_monitorenter) {
+      deoptee_thread->lock_stack().verify_consistent_lock_order(lock_order, exec_mode != Deoptimization::Unpack_none);
+    }
```
The above checks would also hit the follow false positives:
1. deopt in counter overflow in prologue, not in monitorenter
2. monitorenter at bci 0 when raw_bci is -1 (assuming it got past the verifier)
but seems mostly harmless to skip checks in those cases.

pchilano replied:
I thought the original check was fine. Could you elaborate on these 2 cases, I didn't really get them.

dean-long replied:
Deoptimization is already expensive, and this edge case is rare, so I think it would be better to compute the actual previous bytecode here, and not use bci - 1.

dean-long replied:
Other debug checks in deoptimization compute oop maps, which have to iterate all the bytecodes, so doing it here also wouldn't be so bad.  Or how about not checking bytecodes and instead checking a flag on JavaThread that says we are in monitor enter native code?

xmas92 replied:
> Deoptimization is already expensive, and this edge case is rare, so I think it would be better to compute the actual previous bytecode here, and not use bci - 1.

Alright I will give that a go then, unless we think the second option is more appropriate. 

> Or how about not checking bytecodes and instead checking a flag on JavaThread that says we are in monitor enter native code?

What state would that be? Any current state we setup is after transitioning to VM which seems to late. I guess it would be possible to instrument `SharedRuntime::monitor_enter_helper` with some thread local state we can check. But the interpreter has its own entry point. But I guess we would never reach this point for the interpreter, as there is no such thing as deoptimizing or unpacking interpreted frames. 


What is the feeling here? What would be more appropriate? Adding some thread local debug state that we set before transitioning to VM in `SharedRuntime::monitor_enter_helper`  seems like the most precise solution, but we need to be sure that there are no earlier safepoint polls before this point within the execution of the monitorenter bytecode.

xmas92 replied:
I see there is `_pending_monitorenter` but this would only handle synchronized method entry.

dean-long replied:
I like the idea of a flag better, because it is foolproof.  Why can't we set it in ObjectSynchronizer::enter?  I don't think it matters if there is a safepoint check before that, because the lock stack is still consistent at that point.

xmas92 replied:
(Currently) the lock stack is always consistent at safepoints w.r.t. what is actually locked. However the lock stack may not be consistent with the most recent lock returned by the leaf `compiledVFrame::monitors()`. 

But you are correct that it can probably be moved to `ObjectSynchronizer::enter` there are no safepoint polls between `SharedRuntime::monitor_enter_helper` and that point. Similarly there are no safepoints polls in the runtime until after `set_current_pending_monitor` is called. 

So with these following assumptions.
1. LockStack is consistent at safepoints w.r.t. locked monitors
2. No safepoint polls exist from the point that compiledVFrame::monitors() starts returning the monitorinfo for the currently executing monitorenter until either it calls into the runtime or finishes locking.

I do not believe 1. is likely to ever change. But I have limited understanding of the validity of 2. nor if it something that can change.

If both these assumptions are correct than simply skipping the verification when `deoptee->current_pending_monitor() != nullptr` would suffice.

dean-long replied:
I believe those assumptions will always hold, but per separate discussions, let's go with what you have.

dcubed-ojdk commented:
Please clarify what pre-integration testing is being done. As far as I can
tell, this failure only shows up in Tier8 so that should be part of your mix.

dcubed-ojdk commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -440,8 +440,19 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
+    const Bytecodes::Code bc = chunk->first()->method()->code_at(bci);
+    const bool previous_bc_not_monitorenter = bci == 0 ? !chunk->first()->method()->is_synchronized()
+                                                       : chunk->first()->method()->code_at(bci - 1) != Bytecodes::_monitorenter;
+    const bool is_syncronized_entry = chunk->first()->method()->is_synchronized() &&
+                                      chunk->first()->raw_bci() == SynchronizationEntryBCI;
```
nit typo: s/is_syncronized_entry/is_synchronized_entry/

dcubed-ojdk commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -440,8 +440,19 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
+                                                       : chunk->first()->method()->code_at(bci - 1) != Bytecodes::_monitorenter;
+    const bool is_syncronized_entry = chunk->first()->method()->is_synchronized() &&
+                                      chunk->first()->raw_bci() == SynchronizationEntryBCI;
+    // If deoptimizing from monitorenter bytecode we maybe in transitional state. Skip verification.
```
nit typo: s/we maybe/we may be/
nit typo: s/in transitional state/is a transitional state/

dcubed-ojdk commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -440,8 +440,19 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
     }
   }
 #ifdef ASSERT
-  if (LockingMode == LM_LIGHTWEIGHT && !realloc_failures) {
```
In the new code, you are no longer account for `realloc_failures` being true.
I'm not convinced that is okay here.

xmas92 replied:
Originally I had those conditions in as well (to make it more clear and explicit). But Vladimir thought them superfluous. They are implicitly true from `lock_order.is_nonempty() -> LockingMode == LM_LIGHTWEIGHT && !realloc_failures` because we only ever add elements to `lock_order` if `LockingMode == LM_LIGHTWEIGHT && !realloc_failures` is true.

dcubed-ojdk replied:
Thanks for the clarification.

dcubed-ojdk commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -440,8 +440,19 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
+    const bool is_syncronized_entry = chunk->first()->method()->is_synchronized() &&
+                                      chunk->first()->raw_bci() == SynchronizationEntryBCI;
+    // If deoptimizing from monitorenter bytecode we maybe in transitional state. Skip verification.
+    // When reexecuting the current bc, the previous bc may not have finished yet.
```
Should this:
`... the previous bc may not have finished yet.`
be:
`... the previous monitorenter bc may not have finished yet.`

xmas92 replied:
Yeah, while both statements are true, the later is more specific to this context. Which  is probably better and more clear.

dcubed-ojdk reviewed (approved):
Thumbs up on the changes themselves.

Unless I missed it, I haven't seen updated info on the pre-integration testing
in general and specifically about whether Tier8 was executed.

pchilano commented:
I was playing with a reproducer, maybe it would be good to add it: https://github.com/pchilano/jdk/commit/77a85a0ea0dd650929799d8546322d31b69f92e6

xmas92 commented:
> Thumbs up on the changes themselves.
> 
> Unless I missed it, I haven't seen updated info on the pre-integration testing in general and specifically about whether Tier8 was executed.

Just finished Tier8. Also stress tested the reproducer (both @pchilano and the reproducers from Tier8)

> I was playing with a reproducer, maybe it would be good to add it: [pchilano@77a85a0](https://github.com/pchilano/jdk/commit/77a85a0ea0dd650929799d8546322d31b69f92e6)

Great thanks, I'll add it.

/contributor add @pchilano

xmas92 commented:
There were some additional issues with 10d70ea that has to be resolved before this could go in. And multiple engineers in this area agree that the bytecode filter  is probably not the way to solve this. [JDK-8331307](https://bugs.openjdk.org/browse/JDK-8331307) has been created to redo `verify_consistent_lock_order`. 

Because this issue is creating false positives in testing I propose that we first remove the verification so it can be redone. 

Reverted the test and removed the verification logic.

xmas92 commented:
Thanks for the reviews.

/integrate
